Solutions/Corvus.Azure.Storage.Tenancy/Microsoft/Extensions/DependencyInjection/TenancyTableStorageServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,61 @@
+﻿// <copyright file="TenancyTableStorageServiceCollectionExtensions.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    using System;
+    using System.Linq;
+    using Corvus.Azure.Storage.Tenancy;
+
+    /// <summary>
+    /// Common configuration code for services with stores implemented on top of tenanted
+    /// storage.
+    /// </summary>
+    public static class TenancyTableStorageServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Adds services required by tenancy Azure storage based stores.
+        /// </summary>
+        /// <param name="services">The service collection.</param>
+        /// <param name="options">Configuration for the TenantCloudTableFactory.</param>
+        /// <returns>The modified service collection.</returns>
+        public static IServiceCollection AddTenantCloudTableFactory(
+            this IServiceCollection services,
+            TenantCloudTableFactoryOptions options)
+        {
+            return services.AddTenantCloudTableFactory(_ => options);
+        }
+
+        /// <summary>
+        /// Add components for constructing tenant-specific table storage containers.
+        /// </summary>
+        /// <param name="services">The target service collection.</param>
+        /// <param name="getOptions">Function to get the configuration options.</param>
+        /// <returns>The service collection.</returns>
+        public static IServiceCollection AddTenantCloudTableFactory(
+            this IServiceCollection services,
+            Func<IServiceProvider, TenantCloudTableFactoryOptions> getOptions)
+        {
+            if (services is null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (services.Any(s => typeof(ITenantCloudTableFactory).IsAssignableFrom(s.ServiceType)))
+            {
+                return services;
```

Comment:
I wonder if we should throw an exception here.

This pattern of bailing out when the services in question are already registered is fine in cases where you're just saying that you want to enable the service, but here, the caller has supplied options. If we're going to completely ignore those options because someone else already configured this, should we really just do that silently?

Reply:
Good point. This is a common pattern we've used across a lot of our Corvus and Marain libraries, so I'm not going to unilaterally change it here but it's worth further discussion.